snowflake/connect.go

```diff
@@ -0,0 +1,280 @@
+package snowflake
+
+import (
+	"context"
+	"crypto/rsa"
+	"database/sql"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	homedir "github.com/mitchellh/go-homedir"
+	"github.com/pkg/errors"
+	gosnowflake "github.com/snowflakedb/gosnowflake"
+	"github.com/youmark/pkcs8"
+	"golang.org/x/crypto/ssh"
+
+	"github.com/turbot/steampipe-plugin-sdk/v2/plugin"
+)
+
+func connect(ctx context.Context, d *plugin.QueryData) (*sql.DB, error) {
+
+	// have we already created and cached the session?
+	cacheKey := "snowflake"
+	if cachedData, ok := d.ConnectionManager.Cache.Get(cacheKey); ok {
+		return cachedData.(*sql.DB), nil
+	}
+
+	config := GetConfig(d.Connection)
+	var account, user, password, oauthAccessToken, privateKeyPath, privateKey, privateKeyPassphrase, region, role, oauthEndpoint, oauthClientSecret, oauthClientID, oauthRefreshToken, oauthRedirectURL string
+	var browserAuth bool
+	if config.Account != nil {
+		account = *config.Account
+	}
+	if config.User != nil {
+		user = *config.User
+	}
+	if config.Password != nil {
+		password = *config.Password
+	}
+	if config.BrowserAuth != nil {
+		browserAuth = *config.BrowserAuth
+	}
+	if config.PrivateKeyPath != nil {
+		privateKeyPath = *config.PrivateKeyPath
+	}
+	if config.PrivateKey != nil {
+		privateKey = *config.PrivateKey
+	}
+	if config.PrivateKeyPassphrase != nil {
+		privateKeyPassphrase = *config.PrivateKeyPassphrase
+	}
+	if config.PrivateKeyPassphrase != nil {
+		privateKeyPassphrase = *config.PrivateKeyPassphrase
+	}
+	if config.Region != nil {
+		region = *config.Region
+	}
+	if config.Role != nil {
+		role = *config.Role
+	}
+	if config.OAuthEndpoint != nil {
+		oauthEndpoint = *config.OAuthEndpoint
+	}
+	if config.OAuthClientSecret != nil {
+		oauthClientSecret = *config.OAuthClientSecret
+	}
+	if config.OAuthClientID != nil {
+		oauthClientID = *config.OAuthClientID
+	}
+	if config.OAuthRefreshToken != nil {
+		oauthRefreshToken = *config.OAuthRefreshToken
+	}
+	if config.OAuthRedirectURL != nil {
+		oauthRedirectURL = *config.OAuthRedirectURL
+	}
+	if config.OAuthAccessToken != nil {
+		oauthAccessToken = *config.OAuthAccessToken
+	}
+
+	if config.OAuthRefreshToken != nil {
+		accessToken, err := GetOauthAccessToken(oauthEndpoint, oauthClientID, oauthClientSecret, GetOauthData(oauthRefreshToken, oauthRedirectURL))
+		if err != nil {
+			return nil, fmt.Errorf("could not retreive access token from refresh token. Error: %w", err)
```

Comment:
```suggestion
			return nil, fmt.Errorf("could not retreive access token from refresh token: %w", err)
```

For all error messages, we can shorten by removing the word `Error`